### PR TITLE
AzureBlobStorage reporting successful upload when missing credentials

### PR DESCRIPTION
**Pull request by taras**
## Motivation

I working on integrating AzureBlobStorage Publisher into [techdocs-cli](https://github.com/backstage/techdocs-cli/pull/28). In the process, I discovered that AzureBlobStorage is reporting that upload was successful even when it doesn't have credentials. 

## Approach

1. Improved `@azure/storage-blob` mocks to return a response that matches TypeScript definition
2. Removed unnecessary capturing of promises into an array
3. `await promises.push(Promise.resolve(10))` will resolve to 1 which not 10, because it'll resolve to return value of `array.push`
4. Fixed type definition of exists
5. Added tests for cases when files couldn't upload and container name is wrong

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4539-wokjc4hpkbn5w6yv4ljtvmmzvtpizt4q.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602aae00f285af35558f18dc

**Comment by OrkoHunter** [comment]
Thank you for raising this mix of issue and PR @taras ! 😄 

https://github.com/backstage/backstage/blob/29fbdc7e9a86548d562c1c3d04442d580a834b93/packages/techdocs-common/src/stages/publish/azureBlobStorage.ts#L144-L147

Looking at the type of the response to this, it looks like the Promise response has an HTTP Status code
https://github.com/Azure/azure-sdk-for-js/blob/bc9154fd59e8564dea54ad73993bd9aa2ec9a416/sdk/storage/storage-blob/src/Clients.ts#L4897-L4939

My initial guess is that the promise is resolved in all cases but the HTTP Status code might not be 200.

**Comment by taras** [comment]
> My initial guess is that the promise is resolved in all cases but the HTTP Status code might not be 200.

I thought the same, but it's stranger than that. (uploadFile(filePath))[https://github.com/backstage/backstage/blob/29fbdc7e9a86548d562c1c3d04442d580a834b93/packages/techdocs-common/src/stages/publish/azureBlobStorage.ts#L147] resolves to an empty string.

**Comment by taras** [comment]
Ok, well the mystery of empty strings is solved. It's being mocked out. https://github.com/backstage/backstage/blob/c2dfd1cf0135ddef22a1dc647a0b7e75cc6e133d/packages/techdocs-common/__mocks__/%40azure/storage-blob.ts#L26

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4539-wokjc4hpkbn5w6yv4ljtvmmzvtpizt4q.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602aae00f285af35558f18dc

**Comment by freben** [comment]
Hm it complains about prettier in the changeset file.

File: .changeset/quiet-dots-mix.md
Comment (OrkoHunter) [inline .changeset/quiet-dots-mix.md:2]: Prettier
```suggestion
'@backstage/techdocs-common': patch
```

File: packages/techdocs-common/src/stages/publish/azureBlobStorage.ts
Comment (OrkoHunter) [inline packages/techdocs-common/src/stages/publish/azureBlobStorage.ts:89]: Let's throw an error from this block. We don't want to initialize an Azure publisher which is pretty much useless without Auth 😅

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4539-wokjc4hpkbn5w6yv4ljtvmmzvtpizt4q.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602aae00f285af35558f18dc

File: packages/techdocs-common/src/stages/publish/azureBlobStorage.ts
Comment (OrkoHunter) [inline packages/techdocs-common/src/stages/publish/azureBlobStorage.ts:88]: Could we also log the status code and message if any?

File: packages/techdocs-common/src/stages/publish/azureBlobStorage.ts
Comment (OrkoHunter) [inline packages/techdocs-common/src/stages/publish/azureBlobStorage.ts:170]: We should keep the `throw new Error()` to make sure the promise returned by `publish` is rejected.

File: packages/techdocs-common/src/stages/publish/azureBlobStorage.ts
Comment (OrkoHunter) [inline packages/techdocs-common/src/stages/publish/azureBlobStorage.ts:165]: We should throw error in this block - or else the promise will be resolved.

**Review by OrkoHunter — APPROVED** [review]
Nice! Thanks for fixing this.

File: packages/techdocs-common/src/stages/publish/azureBlobStorage.ts
Comment (taras) [reply packages/techdocs-common/src/stages/publish/azureBlobStorage.ts]: Done in https://github.com/backstage/backstage/pull/4539/commits/4de00ead3887a1dd8307745b1ab46d3e40a1f58a

File: packages/techdocs-common/src/stages/publish/azureBlobStorage.ts
Comment (taras) [reply packages/techdocs-common/src/stages/publish/azureBlobStorage.ts]: Done in https://github.com/backstage/backstage/pull/4539/commits/4de00ead3887a1dd8307745b1ab46d3e40a1f58a

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4539-wokjc4hpkbn5w6yv4ljtvmmzvtpizt4q.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602aae00f285af35558f18dc

File: packages/techdocs-common/src/stages/publish/azureBlobStorage.ts
Comment (taras) [reply packages/techdocs-common/src/stages/publish/azureBlobStorage.ts]: Done in https://github.com/backstage/backstage/pull/4539/commits/02c5ce979c49aac15b6e4dde159905598d75e79a

**Comment by OrkoHunter** [comment]
`yarn run prettier --write .changeset/*.md` :)

LGTM - either we wait for @backstage/maintainers to review or we update the changeset to be prefixed with `techdocs-` to [utilize the code owners rule](https://github.com/backstage/backstage/blob/8c8db18d3c56dfd384ddb2abe150ce2248e7baea/.github/CODEOWNERS#L18).

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4539-wokjc4hpkbn5w6yv4ljtvmmzvtpizt4q.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602aae00f285af35558f18dc

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4539-wokjc4hpkbn5w6yv4ljtvmmzvtpizt4q.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/602aae00f285af35558f18dc

**Comment by OrkoHunter** [comment]
Since you're testing it as of this moment - let me know when this is good to go.

**Comment by taras** [comment]
@OrkoHunter I manually verified that this is working using a build in https://github.com/backstage/techdocs-cli/pull/28

**Comment by OrkoHunter** [comment]
This is super! ⭐

**Pull request merged**
